### PR TITLE
More mouse improvements

### LINKED ISSUE #821: Improved cursor mode
There could be three cursor modes:
* CURSOR_SYSTEM - Sends coordinates, shows the system cursor
* CURSOR_CUSTOM - Sends coordinates, hides the system cursor
* DELTAS - Sends deltas, hides the cursor

This will make ForceCursor redundant, its calls should be replaced and moved away from the frame loop and the currently present checks should become events.

### PR DESCRIPTION
**Pull request by mbasaglia**
### Three mouse modes
* System cursor: The mouse is handled by the system (movement and cursor)
* Custom cursor: The mouse is free to move and rendered by the game
* Deltas: The mouse is hidden and can't move, input is sent as movement deltas

### Mouse focus engine event
A new input event is sent from the engine to the client to signify that the game area has lost or gained focus. This can happen when the window has lost focus, the console is opened and similar situations.

### All mouse modes are handled by cgame
With the new event and mouse modes, is up to the cgame vm to decide what to do with the cursor, the engine never forces the cursor to be in any particular state

## Bug Fixes
Since the new behaviour is more flexible and cgame has all the information it needs, the mouse is no longer locked on map load.

This merge request closes #821 and #862.

## Remaining issues
<del>The loading screen locks all input event so the focus status has to be evaluated at each frame (instead of reacting from SDL events that change it).
Once the loading screen is over, the mouse will jump to the center of the Unvanquished window (without locking), this is caused by the fact that it's impossible to know what happened to the mouse when the loading screen was running.</del>

### CODE REVIEW

**Comment by DolceTriade** [comment]
In general, the code looks ok to me. However, Mouse still gets stuck in middle of screen when I switch away during map load.

File: daemon/src/engine/null/null_input.cpp [hunk 1]
```diff
@@ -28,6 +28,7 @@ If you have questions concerning this license or the applicable additional terms
 
 #include "qcommon/q_shared.h"
 #include "qcommon/qcommon.h"
+#include "client/client.h"
```
Comment (DolceTriade) [inline daemon/src/engine/null/null_input.cpp]: Why is this include file required?

File: daemon/src/engine/sys/sdl_input.cpp [hunk 2]
```diff
@@ -561,31 +561,56 @@ static keyNum_t IN_TranslateSDLToQ3Key( SDL_Keysym *keysym, bool down )
...
+    return mouse_mode;
+}
 
 /*
- * Activates the cursor, the system cursor will still be disabled and
- * the game is responsible of drawing the cursor image
+ * Enables or disables the cursor
  */
-void IN_SetCursorActive( bool active )
+void IN_SetMouseMode(MouseMode new_mode)
```
Comment (DolceTriade) [inline daemon/src/engine/sys/sdl_input.cpp]: Perfer camelcase for variable names for consistency with the other variables in the file.

**Comment by DolceTriade** [comment]
Ok, turns out the mouse getting stuck is an edge case. If you start a map when the game doesn't have focus (via your tty terminal console, or even through the unvanquished IPC socket, it will get stuck.

File: daemon/src/engine/null/null_input.cpp (hunk 1 above)
Comment (mbasaglia) [inline daemon/src/engine/null/null_input.cpp]: For the MouseMode enum

**Comment by mbasaglia** [comment]
Now everything seems to be working fine.

**Comment by DolceTriade** [comment]
Awesome. LGTM.

**Comment by Kangz** [comment]
LGTM too

**Comment by mbasaglia** [comment]
Merged in for-0.48.0

**Pull request closed without merging**